src/Qir/Runtime/lib/QIR/callables.cpp

```diff
@@ -102,7 +102,11 @@ extern "C"
         {
             for (int i = increment; i < 0; i++)
             {
-                (void)callable->Release();
+                if (0 == callable->Release())
+                {
+                    assert(-1 == i && "Attempting to decrement reference count below zero!");

```

Comment:
nit: I'm not sure I like this pattern, though honestly there isn't anything functionally wrong here. It just seems odd that this assert message would show up only if the reference count was updated with a decrement value less than -1, where if the decrement is exactly -1 (our compiler's QIR default) on an item of 0 references then the assert from within `Release` will fire instead. That assert already protected us from decrementing below zero, so this new one is arguably extraneous, though not harmful.

Reply:
I'm totally with you on our scenario.

But looks like the static analyzer has considered a different scenario and complained for a possible released (dangling) pointer dereference. I'll try to tell what I remember.

Let's consider a `callable` with `callable->refCount` equal `1`. And someone calls `__quantum__rt__callable_update_reference_count()` with `increment` parameter `-2`. During the first most iteration of this `for` loop (`i == -2`) the `callable->Release()` is invoked. It decrements the `callable->refCount` to `0`, deallocates the `callable`, and returns `0`. After the first call to `callable->Release()` the `callable` becomes a dangling pointer.

Before this change, the value `0` returned by `callable->Release()` was not analyzed, and another iteration of the loop was starting. At the moment of the second call `callable->Release()` a dangling pointer was accessed which immediately resulted in the undefined behavior (UB), before the execution reached an assert inside of `QirCallable::Release()`.

I agreed with the complaint from the static analyzer and tried to prevent the UB with an assert (in green line 107). I don't state that my change is the best fix (e.g. the `break` in the green line 108 can be redundant ;-) ), feel free to propose a better fix.

Reply:
Ah, that's a great point! I had incorrectly assumed `Release` would clean up internal structures but not invalidate the `callable` pointer itself, but looking at the implementation that is not the case. Score one for the analyzer ;)